cpp/src/arrow/compute/kernels/scalar_temporal.cc

```diff
@@ -136,19 +136,25 @@ struct Day {
 
 // ----------------------------------------------------------------------
 // Extract day of week from timestamp
+//
+// By default week starts on Monday represented by 0 and ends on Sunday represented
+// by 6. Start day of the week (Monday=1, Sunday=7) and numbering start (0 or 1) can be
+// set using DayOfWeekOptions
 
 template <typename Duration>
 struct DayOfWeek {
-  explicit DayOfWeek(TemporalComponentExtractionOptions options) : options(options) {}
+  explicit DayOfWeek(DayOfWeekOptions options) : options(options) {}
 
   template <typename T, typename Arg0>
   T Call(KernelContext*, Arg0 arg, Status*) const {
-    return static_cast<T>(
-        weekday(year_month_day(floor<days>(sys_time<Duration>(Duration{arg}))))
-            .iso_encoding() -
-        1 + options.week_start);
+    auto wd = arrow_vendored::date::year_month_weekday(
+                  floor<days>(sys_time<Duration>(Duration{arg})))
+                  .weekday()
+                  .iso_encoding();
+    return (wd + 7 - options.week_start) % 7 + options.one_based_numbering;
```

Comment:
Would it be worth to have a separate if branch for the case that `options.week_start` is 1 (the default), in which case this is this is much simpler (eg you don't need the `%`)? 
(I don't know how expensive those calculations are compared to the actual `year_month_weekday(..).weekday()` call, though, so maybe it doesn't matter)

Reply:
I'm not sure about this  either. Added a `options.week_start == 1` branch.

Reply:
If we are worried about the modulo cost, we could instead use a simple condition:
```c++
const auto shifted_wd = wd + 7 - options.week_start;
return options.one_based_numbering + ((shifted_wd > 7) ? shifted_wd - 7 : shifted_wd);
```

Reply:
Or, of course, a simple lookup table could be created in the kernel constructor.

Reply:
This is on my TODO, will get back to it asap.

Reply:
Done.